server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutes.java

```diff
@@ -96,42 +124,68 @@ public void define(Service service) {
+            value = "Compulsory. Needs to ve a valid action represent for an operation to perform on the Deleted Message Vault")
     })
     @ApiResponses(value = {
         @ApiResponse(code = HttpStatus.CREATED_201, message = "Task is created", response = TaskIdDto.class),
         @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - user param is invalid"),
         @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side.")
     })
     private TaskIdDto restore(Request request, Response response) throws JsonExtractException {
+        validateActionQuery(request, Action.RESTORE);
+
         User userToRestore = extractUser(request);
         QueryDTO queryDTO = jsonExtractor.parse(request.body());
         Query query = translate(queryDTO);
         TaskId taskId = taskManager.submit(new DeletedMessagesVaultRestoreTask(vaultRestore, userToRestore, query, queryDTO));
         return TaskIdDto.respond(response, taskId);
     }
 
+    private void validateActionQuery(Request request, Action appropriateAction) {
+        try {
+            assertRequestedAction(request, appropriateAction);
+        } catch (IllegalArgumentException e) {
+            haltError(e);
+        }
+    }
+
+    private void assertRequestedAction(Request request, Action appropriateAction) {
+        String actionParam = request.queryParams(ACTION_QUERY_PARAM);
+        Action requestedAction = Action.getAction(actionParam);
+        Preconditions.checkArgument(appropriateAction.equals(requestedAction),
+            "requestedAction " + requestedAction.name().toLowerCase() + " is not allowed(" + appropriateAction.name().toLowerCase() + ")");
+    }
+
     private Query translate(QueryDTO queryDTO) {
         try {
             return queryTranslator.parse(queryDTO);
         } catch (QueryTranslator.QueryTranslatorException e) {
-            throw ErrorResponder.builder()
-                .statusCode(HttpStatus.BAD_REQUEST_400)
-                .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
-                .message(e.getMessage())
-                .haltError();
+            haltError(e);
         }
+        return null;
     }
 
     private User extractUser(Request request) {
         try {
             return User.fromUsername(request.params(USER_PATH_PARAM));
         } catch (IllegalArgumentException e) {
-            throw ErrorResponder.builder()
-                .statusCode(HttpStatus.BAD_REQUEST_400)
-                .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
-                .message(e.getMessage())
-                .haltError();
+            haltError(e);
         }
+        return null;
+    }
+
+    private void haltError(Exception e) {
+        throw ErrorResponder.builder()
+            .statusCode(HttpStatus.BAD_REQUEST_400)
+            .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+            .message(e.getMessage())
```

Comment:
The exception can be passed as `.cause(e)` which allow passing nice messages like `.message("Error while parsing user")` or `.message("Invalid action query parameter")` or `.message("Error while parsing query body")` which then makesthis ErrorResponder a bit less boring and a bit more useful